syz-verifier/stats.go

```diff
@@ -7,22 +7,133 @@ import (
 	"fmt"
 	"sort"
 	"strings"
+	"sync"
 	"time"
 
 	"github.com/google/syzkaller/prog"
 )
 
+type StatUint64 struct {
+	uint64
+	mu *sync.Mutex
+}
+
+func (s *StatUint64) Inc() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.uint64++
+}
+
+func (s *StatUint64) Get() uint64 {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	return s.uint64
+}
+
+type StatTime struct {
+	time.Time
+	mu *sync.Mutex
+}
+
+func (s *StatTime) Set(t time.Time) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.Time = t
+}
+
+func (s *StatTime) Get() time.Time {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	return s.Time
+}
+
+type mapStringToCallStats map[string]*CallStats
+
+type StatMapStringToCallStats struct {
+	mapStringToCallStats
+	mu *sync.Mutex
+}
+
+func (stat *StatMapStringToCallStats) IncCallOccurrenceCount(key string) {
+	stat.mu.Lock()
+	stat.mapStringToCallStats[key].Occurrences++
+	stat.mu.Unlock()
+}
+
+func (stat *StatMapStringToCallStats) IncMismatches(key string) {
+	stat.mu.Lock()
+	stat.mapStringToCallStats[key].Mismatches++
+	stat.mu.Unlock()
+}
+
+func (stat *StatMapStringToCallStats) AddState(key string, state ReturnState) {
+	stat.mu.Lock()
+	stat.mapStringToCallStats[key].States[state] = true
+	stat.mu.Unlock()
+}
+
+func (stat *StatMapStringToCallStats) SetCallInfo(key string, info *CallStats) {
+	stat.mu.Lock()
+	stat.mapStringToCallStats[key] = info
+	stat.mu.Unlock()
+}
+
+func (stat *StatMapStringToCallStats) totalExecuted() uint64 {
+	stat.mu.Lock()
+	defer stat.mu.Unlock()
+
+	var t uint64
+	for _, cs := range stat.mapStringToCallStats {
+		t += cs.Occurrences
+	}
+	return t
+}
+
+func (stat *StatMapStringToCallStats) orderedStats() []*CallStats {
+	stat.mu.Lock()
+	defer stat.mu.Unlock()
+
+	css := make([]*CallStats, 0)
+	for _, cs := range stat.mapStringToCallStats {
+		if cs.Mismatches > 0 {
+			css = append(css, cs)
+		}
+	}
+
+	sort.Slice(css, func(i, j int) bool {
+		return getPercentage(css[i].Mismatches, css[i].Occurrences) > getPercentage(css[j].Mismatches, css[j].Occurrences)
+	})
+
+	return css
+}
+
 // Stats encapsulates data for creating statistics related to the results
 // of the verification process.
 type Stats struct {
+	mu                  sync.Mutex
+	TotalCallMismatches StatUint64
+	TotalProgs          StatUint64
+	ExecErrorProgs      StatUint64
+	FlakyProgs          StatUint64
+	MismatchingProgs    StatUint64
+	StartTime           StatTime
 	// Calls stores statistics for all supported system calls.
-	Calls               map[string]*CallStats
-	TotalCallMismatches int64
-	TotalProgs          int64
-	ExecErrorProgs      int64
-	FlakyProgs          int64
-	MismatchingProgs    int64
-	StartTime           time.Time
+	Calls StatMapStringToCallStats
+}
+
+func (stats *Stats) Init() *Stats {
+	stats.TotalCallMismatches.mu = &stats.mu
+	stats.TotalProgs.mu = &stats.mu
+	stats.ExecErrorProgs.mu = &stats.mu
+	stats.FlakyProgs.mu = &stats.mu
+	stats.MismatchingProgs.mu = &stats.mu
+	stats.StartTime.mu = &stats.mu
+	stats.Calls.mu = &stats.mu
```

Comment:
Why do they all have to use only one mutex?

Reply:
Hm. My logic was next:
1. Serialized access to the Stats object is the simplest way to understand how it works.
2. Atomic operations on the independent fields improve performance, but require you to think about the variables dependency and makes hard multi-var transactions in the future once they'll be needed.

I believe current implementation is easier to understand. If not - please let me know.